lib/openstack_api/openstack_image.py

```diff
@@ -284,11 +287,55 @@ def find_non_existent_projects(self, cloud_account: str) -> Dict[str, List[str]]
         :param cloud_account: The associated clouds.yaml account
         :return: A dictionary containing the non-existent projects and a list of images that refer to them
         """
-        return self._query_api.find_non_existant_object_projects(
+        return self._query_api.find_non_existent_object_projects(
             cloud_account=cloud_account,
             check_params=NonExistentProjectCheckParams(
                 object_list_func=lambda conn: conn.list_images(),
                 object_id_param_name="id",
                 object_project_param_name="owner",
             ),
         )
+
+    # pylint:disable=too-many-arguments
+    def email_users(
```

Comment:
I think we should extract this method into a parent class, something like:
```
from abc import ABC
class EmailAction(ABC):
   def email_users(...):
        ...

   def email_support(...):
        ...
```

Then this class gets mixed into the actions:
```
class OpenstackImage(OpenstackWrapperBase, EmailAction):
```

This would give us a completely consistent set of params and remove the duplication
Thoughts?

Reply:
The only complication I can see is needing an OpenstackQuery instance in the new class as well which could be used in OpenstackImage but creates a difference in code between resources that have email actions and those that only have a list one. I do think the other refactor for the querying could do with a base class for handling the search functions however. If I do that over there then I could have new 'EmailAction' inherit from that to get the openstack query instance as well as implement these functions which would solve it?

Reply:
So I would break up the action code so that the OpenstackQuery only has the email step, and the query happens in the action

Something like:

```
def query_and_email():
    results = query()
    super().email_users(results)
```
If that's possible?

Reply:
There is also the additional check of whether the parameters given to the query are valid e.g. if they need a project or not, but this check is only needed for emails and not the list actions. Also I thought the idea was we wanted a single function that could run the query and email in the lib folder and not rely on the actions?

Reply:
So I'm happy for the actions layer to call two methods from the lib layer, as long as the implementation isn't in the actions layer

It's just separation of concerns is all